packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,406 @@
+				byte_offset bigint(20) unsigned NOT NULL,
+				sort_order int DEFAULT 1,
+				PRIMARY KEY  (id),
+				KEY session_id (session_id),
+				KEY entity_id (entity_id(%d)),
+				KEY parent_id (parent_id(%d)),
+				KEY byte_offset (byte_offset)
+			) ' . $wpdb->get_charset_collate(),
+			self::get_table_name(),
+			$max_index_length,
+			$max_index_length
+		);
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		dbDelta( $sql );
+	}
+
+	/**
+	 * Run by register_deactivation_hook. It drops the table and deletes the
+	 * option.
+	 */
+	public static function deactivate() {
+		global $wpdb;
+		$table_name = self::get_table_name();
+
+		// Drop the table.
+		$wpdb->query( $wpdb->prepare( 'DROP TABLE IF EXISTS %s', $table_name ) );
+
+		// Delete the option.
+		delete_option( self::OPTION_NAME );
+	}
+
+	/**
+	 * Reset the class.
+	 */
+	public function reset() {
+		$this->set_session( null );
+	}
+
+	/**
+	 * Set the current session ID.
+	 *
+	 * @param int|null $session_id The session ID.
+	 */
+	public function set_session( $session_id ) {
+		$this->current_session = $session_id;
+	}
+
+	/**
+	 * Delete all rows for a given session ID.
+	 *
+	 * @param int $session_id The session ID to delete rows for.
+	 * @return int|false The number of rows deleted, or false on error.
+	 */
+	public function delete_session( $session_id = null ) {
+		global $wpdb;
+
+		return $wpdb->delete(
+			self::get_table_name(),
+			array( 'session_id' => $session_id ?? $this->current_session ),
+			array( '%d' )
+		);
+	}
+
+	/**
+	 * Map an entity to the index. If $id is provided, it will be used to map the entity.
+	 *
+	 * @param string   $entity_type The type of the entity.
+	 * @param array    $data The data to map.
+	 * @param int|null $id The ID of the entity.
+	 * @param int|null $additional_id The additional ID of the entity.
+	 */
+	public function map_entity( $entity_type, $data, $id = null, $additional_id = null ) {
+		global $wpdb;
+
+		if ( ! array_key_exists( $entity_type, self::ENTITY_TYPES ) ) {
+			return;
+		}
+
+		$new_entity = array(
+			'session_id'   => $this->current_session,
+			'entity_type' => self::ENTITY_TYPES[ $entity_type ],
+			'entity_id'   => null,
+			'mapped_id'    => is_null( $id ) ? null : (string) $id,
+			'parent_id'    => null,
+			'byte_offset'  => 0,
+			// Items with a parent has at least a sort order of 2.
+			'sort_order'   => 1,
+		);
+		// Get the ID of the entity.
+		$entity_id = (string) $data[ self::ENTITY_TYPES_ID[ $entity_type ] ];
+
+		// Map the parent ID if the entity has one.
+		switch ( $entity_type ) {
+			// @TODO: missing comment parent ID.
+			case 'comment_meta':
+				if ( array_key_exists( 'comment_id', $data ) ) {
+					$new_entity['parent_id'] = $data['comment_id'];
+				}
+				break;
+			case 'post':
+				if ( 'post' === $data['post_type'] || 'page' === $data['post_type'] ) {
+					if ( array_key_exists( 'post_parent', $data ) && '0' !== $data['post_parent'] ) {
```

Comment:
we're dealing with raw user input here so we may also need to check if it's an int etc